core/src/Revolution/modX.php

```diff
@@ -407,7 +407,7 @@ public static function toQueryString(array $parameters = [], $numPrefix = '', $a
      * @throws xPDOException
      */
     public static function getInstance($id = null, $config = null, $forceNew = false) {
-        $class = __CLASS__;
+        $class = class_exists($id) ? $id : __CLASS__;
```

Comment:
The `$id` here is NOT the class name. It is either a random id or a named key you provide. In the default case, it should be left null. This will ALWAYS return the existing, single instance that was first instantiated in the current request, as long as you do not also pass anything other than NULL to the `$config` parameter. In the case of connectors/index.php and manager/index.php, it is fine that they pass the additional `$config` parameter.

Reply:
Can revert this change to the `$class` variable here?

Reply:
> Can revert this change to the `$class` variable here?

So far I see that it will work the same
But if I want to use my own modx class, for example modY, is there an option with $class = class_exists($id)? $id : __CLASS__;

It will work better, where I will only need to specify the class name in 3 index.php files

There is a choice, but the working option is both.

Reply:
For a better understanding of how getInstance will work
I agree, the old option is more suitable
To avoid surprises and to make the entry point into the application clear

Reply:
submitted changes

Reply:
in modx 2.8 it turns out you can make similar changes to index.php $modx = modX::getInstance();

Reply:
If you want to be able to extend modX and have it use the actual classname of your extended class to generate the uniqid prefix (which is all this is doing), we can change that to a late static binding, e.g. `static::class`.

Reply:
> If you want to be able to extend modX and have it use the actual classname of your extended class to generate the uniqid prefix (which is all this is doing), we can change that to a late static binding, e.g. `static::class`.


No)))
The fact is that the experience of using this approach was precisely in the call $modx = modY::getInstance('modY');
Based on this experience

The option that is now in the pull request is what we need.
The main idea is to call an instance in one tab via Live template (PhpStorm)
$modx = modY::getInstance('modY');
and start working from any class
with code highlighting ($modx->) immediately without any comments or other additional code.

Reply:
The code highlighting works from the class itself, not from the string you pass to the function. I'm lost on what you are talking about here. Regardless of what string exists as the class, the unique id that is generated with that string as the prefix is what defines the instance that will be retrieved on additional calls to the method when you DO NOT PROVIDE an id. If you provide an id, the class variable is irrelevant. But, if you call getInstance from a class modY that extends modX, class is relevant. This is the ONLY time it is relevant, and in that case, it needs to use late static binding to automatically set the class that is instantiated regardless of what the `$id` is.

Reply:
![CleanShot 2024-05-15 at 00 52 21](https://github.com/modxcms/revolution/assets/5019138/44719771-929b-48bb-9c8a-1dd70eb8c07b)

I'm talking about modifying config.core.php
if I connect in config.core.php autoloader.php from composer
then when called it will extend modX and then I can access it using

$modx = modY::getInstance('modY');


I don’t focus on this, this experience allowed me to fully use the namespace "App\\": "src/"
where I never passed public function __construct(modX $modx)

There was always a call: modY::getInstance('modY');

Let me explain, I used the modY class because it was a combat project, and it was a risk. But the risk was justified, everything works fully

You showed that modX can be used without specifying $id

Reply:
Again, you don't need to pass 'modY' to the getInstance() function if we use `static::class` (late static binding) instead of `__CLASS__` to assign the `$class` variable in modX::getInstance()… modY::getInstance() will automatically know it is modY.

Reply:
I can make that adjustment in conjunction with yours if you are okay with me doing that. I will also adjust the return type so IDEs will also know it returns the class from the late static binding.

Reply:
> I can make that adjustment in conjunction with yours if you are okay with me doing that. I will also adjust the return type so IDEs will also know it returns the class from the late static binding.

ok